File: server/src/routes/leaderboard.ts
import { Router } from "express";
import { prisma } from "../lib/prisma.js";
import { leaderboardQuerySchema } from "../lib/contracts.js";

const router = Router();

router.get("/", async (req, res, next) => {
  try {
    const query = leaderboardQuerySchema.parse(req.query);

    const games = await prisma.game.findMany({
      where: {
        mode: query.mode,
      },
      orderBy: [
        { totalScore: "desc" },
        { totalDistanceKm: "asc" },
        { playedAt: "asc" },
      ],
      include: {
        player: true,
      },
      take: query.limit,
    });

    res.json({
      items: games.map((game) => ({
        id: game.id,
        playerName: game.player.name,
        mode: game.mode,
        totalScore: game.totalScore,
        averageScore: game.averageScore,
        totalDistanceKm: game.totalDistanceKm,
        playedAt: game.playedAt.toISOString(),
      })),
    });
  } catch (error) {
    next(error);
  }
});

export default router;

File: app/routers/leaderboard.py
from __future__ import annotations

from fastapi import APIRouter, Depends, Query
from sqlalchemy.ext.asyncio import AsyncSession

from .. import services
from ..database import get_session
from ..models import GameMode
from ..schemas import LeaderboardEntry, LeaderboardResponse

router = APIRouter(prefix="/leaderboard", tags=["leaderboard"])


@router.get(
    "",
    response_model=LeaderboardResponse,
    response_model_by_alias=True,
)
async def get_leaderboard(
    mode: GameMode | None = Query(default=None),
    limit: int = Query(default=20, ge=1, le=100),
    session: AsyncSession = Depends(get_session),
) -> LeaderboardResponse:
    rows = await services.fetch_leaderboard(session, mode, limit)

    items = [
        LeaderboardEntry(
            id=game.id,
            player_name=player.name,
            mode=game.mode,
            total_score=game.total_score,
            average_score=game.average_score,
            total_distance_km=game.total_distance_km,
            played_at=game.played_at,
        )
        for game, player in rows
    ]
    return LeaderboardResponse(items=items)
